### PR TITLE
Highway Layer Porting

### PR DESCRIPTION
**Pull request by Ash-29**
## Description
This PR is to port Highway to PyTorch.
Fix #(issue)
Modified dense layers and added unit test

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/layers.py [hunk 1]
```diff
@@ -3743,3 +3743,71 @@ def forward(
             P = pair_features
 
         return [A, P]
+
+class Highway(nn.Module):
+    """
+        Highway layer used in TextCNN model.
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:3749]: Formatting here is off. Please make sure to check alignment

File: deepchem/models/torch_models/layers.py [hunk 2]
```diff
@@ -3743,3 +3743,71 @@ def forward(
...
+
+        """
+    def __init__(self,
+                 activation: str='relu',
+                 biases_initializer: str='zeros',
+                 weights_initializer='kaiming_uniform_',
+                 layer_shape=[5, 2],
+                 **kwargs):
+        """
+              Parameters
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:3774]: Formatting

File: deepchem/models/torch_models/layers.py [hunk 3]
```diff
@@ -3743,3 +3743,71 @@ def forward(
...
+        self.linear_H = init_func(torch.empty([input_shape, out_channels]))
+        self.linear_T = init_func(torch.empty([input_shape, out_channels]))
+
+    def __repr__(self) -> str:
+        return (
+            f'{self.__class__.__name__}(activation_fn={self.activation_fn}, '
+            f'biases_initializer={self.biases_initializer}, '
+            f'weights_initializer={self.weights_initializer})')
+
+    def forward(self, inputs: torch.Tensor):
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:3804]: Please add a docstring

File: deepchem/models/torch_models/layers.py [hunk 4]
```diff
@@ -3743,3 +3743,71 @@ def forward(
             P = pair_features
 
         return [A, P]
+
+class Highway(nn.Module):
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:3747]: You will need to add this class to the docs/ folder. See where the other layers are and add to the same rst file

File: deepchem/models/torch_models/tests/highway_test.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import torch
+import torch.nn as nn
+import torch.nn.init as init
+
+def test_highway():
```
Comment (rbharath) [inline deepchem/models/torch_models/tests/highway_test.py:5]: You need to add a test that this layer produces the same outputs as the TF layer on the same inputs. See other tests for porting added in recent ports

**Review by rbharath — COMMENTED** [review]
A good start. We need to clean up comments, add docs, and add more tests before this will be ready

**Comment by Ash-29** [comment]
I haven't added linear_T in shapes test, I'll add, just matching the outputs

File: deepchem/models/torch_models/layers.py [hunk 6]
```diff
@@ -3743,3 +3743,71 @@ def forward(
...
+
+        input_shape = layer_shape[0]
+        out_channels = layer_shape[1]
+
+        init_func = getattr(initializers, self.weights_initializer)
+
+        self.linear_H = init_func(torch.empty([input_shape, out_channels]))
+        self.linear_T = init_func(torch.empty([input_shape, out_channels]))
+
+    def __repr__(self) -> str:
```
[the remark this answers is not in the document]
Comment (Ash-29) [reply deepchem/models/torch_models/layers.py:3904]: Done

File: deepchem/models/torch_models/layers.py (hunk 2 above)
Comment (Ash-29) [reply deepchem/models/torch_models/layers.py:3879]: I run yapf and mypy to resolve formatting issues

Comment (rbharath) [reply deepchem/models/torch_models/layers.py:3879]: The docstring is still improperly spaced. See how the comment goes in past the quotes.

File: deepchem/models/torch_models/tests/highway_test.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  output_tf = (5, 3)
+     # Testing Shapes with TF Model Output
+  assert output.shape == output_tf
+  assert dense_H.shape == dense_H
+  assert dense_T.shape == dense_T
+
+     # Testing Shapes
+  assert output.shape == (width,batch_size)
+  assert layer.activation == torch.ReLU
+
```
Comment (rbharath) [inline deepchem/models/torch_models/tests/highway_test.py:46]: You need to add a test here that the model produces exactly the same outputs on exactly the same inputs. See how other recent porting PRs do this (you will need to store .npy files with sample input/output arrays to test)

File: deepchem/models/torch_models/layers.py (hunk 2 above)
Comment (Ash-29) [reply deepchem/models/torch_models/layers.py:3879]: Oh, I see , I'll fix that

File: deepchem/models/torch_models/layers.py (hunk 3 above)
Comment (Ash-29) [reply deepchem/models/torch_models/layers.py:3922]: Done

File: deepchem/models/torch_models/tests/highway_test.py (hunk 7 above)
Comment (Ash-29) [reply deepchem/models/torch_models/tests/highway_test.py:46]: Alright, I'll add .npy files

File: deepchem/models/torch_models/layers.py (hunk 1 above)
Comment (Ash-29) [reply deepchem/models/torch_models/layers.py:3853]: I'll fix the spacing issues, I want to ask for rest formating is yapf and mypy enough or should I do mannual check ?

File: deepchem/models/torch_models/layers.py (hunk 4 above)
Comment (Ash-29) [reply deepchem/models/torch_models/layers.py:3851]: Sure, once the formatting and testing is done then I'll add

File: deepchem/models/torch_models/layers.py (hunk 1 above)
Comment (ARY2260) [reply deepchem/models/torch_models/layers.py:3853]: Consider testing: yapf, flake8, doctest, pytest, mypy for each .py file you modify

File: deepchem/models/torch_models/layers.py (hunk 2 above)
Comment (ARY2260) [reply deepchem/models/torch_models/layers.py:3879]: Have you fixed these?

File: deepchem/models/torch_models/layers.py [hunk 8]
```diff
@@ -3846,3 +3846,103 @@ def forward(
...
+        """Returns a string representing the configuration of the layer.
+
+        Returns
+        -------
+        activation_fn: str, optional
```
Comment (ARY2260) [inline deepchem/models/torch_models/layers.py:3909]: `optional` doesn't make sense in `return ` part of docstrings

File: deepchem/models/torch_models/layers.py [hunk 9]
```diff
@@ -3846,3 +3846,103 @@ def forward(
...
+        """
+        if isinstance(inputs, (list, tuple)):
+            parent = inputs[0]
+        else:
+            parent = inputs
+
+        parent = parent.view(self.linear_H.shape)
+        linear_H = self.activation_fn(self.linear_H * parent)
+        linear_T = torch.sigmoid(self.linear_T * parent)
+        return linear_H * linear_T + parent * (1 - linear_T)
```
Comment (ARY2260) [inline deepchem/models/torch_models/layers.py:3948]: comment the expected shape of output

File: deepchem/models/torch_models/tests/highway_test.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+  width = 5
+  batch_size = 2
+  input = torch.rand(batch_size, width, dtype=torch.float32)
+  layer = Highway(layer_shape=[width, 2])
+  result = layer(input)
+  assert result.shape == (batch_size, width)
+  assert len(layer.parameters()) == 4
+
+  # Creating a second layer should produce different results, since it has
+  # different random weights.
```
Comment (ARY2260) [inline deepchem/models/torch_models/tests/highway_test.py:16]: split this into a separate test and use random seed to avoid randomness while testing

File: deepchem/models/torch_models/tests/highway_test.py [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+        layer.T.weight.data.weight.data = torch.from_numpy(
+         np.transpose(tf_weights['deepchem/models/tests/assets/TF_weights_T.npy']))
+        layer.T.weight.data.bias.data = torch.from_numpy(
+         tf_weights['deepchem/models/tests/assets/TF_bias_T.npy'])   
+         
+    output_tensor = torch.from_numpy(
+         np.load('deepchem/models/tests/assets/highway_output.npy').astype(
+             np.float32))
+  
+    assert torch.allclose(output, output_tensor, atol=1e-04)
```
Comment (ARY2260) [inline deepchem/models/torch_models/tests/highway_test.py:54]: what `output` value are you testing here? the forward pass is missing in the test

**Review by ARY2260 — CHANGES_REQUESTED** [review]
I have reviewed the PR and added some comments. Please check the unit tests locally first before review.

File: deepchem/models/torch_models/layers.py (hunk 8 above)
Comment (Ash-29) [reply deepchem/models/torch_models/layers.py:3909]: I've removed that

File: deepchem/models/torch_models/layers.py (hunk 2 above)
Comment (Ash-29) [reply deepchem/models/torch_models/layers.py]: fixed

File: deepchem/models/torch_models/layers.py (hunk 4 above)
Comment (rbharath) [reply deepchem/models/torch_models/layers.py:3851]: Make sure to also add it to torch_layers.csv in addition to the layers.rst file in the docs. Please add it before next review

File: deepchem/models/torch_models/tests/highway_test.py (hunk 11 above)
Comment (Ash-29) [reply deepchem/models/torch_models/tests/highway_test.py:68]: Done

File: deepchem/models/torch_models/tests/highway_test.py (hunk 10 above)
Comment (Ash-29) [reply deepchem/models/torch_models/tests/highway_test.py:19]: Done

File: deepchem/models/torch_models/layers.py (hunk 4 above)
Comment (Ash-29) [reply deepchem/models/torch_models/layers.py]: Yes, added

File: deepchem/models/tests/highway_test.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
+import torch
+import torch.nn as nn
+import torch.nn.init as init
+import torch.nn.functional as F
+from deepchem.models.torch_models.layers import Highway
+import numpy as np
+
+class Highway(torch.nn.Module):
```
Comment (rbharath) [inline deepchem/models/tests/highway_test.py:8]: We shouldn't have this new class here

File: deepchem/models/tests/highway_test.py [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+  def __init__(self, layer_shape):
+        super(Highway, self).__init__()
+
+  def test_highway_layer1():
```
Comment (rbharath) [inline deepchem/models/tests/highway_test.py:12]: Let's have a more descriptive test name

File: deepchem/models/torch_models/layers.py [hunk 14]
```diff
@@ -4332,3 +4332,112 @@ def gaussian_histogram(self, x: torch.Tensor) -> torch.Tensor:
...
+    >>> input = torch.tensor([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
+    >>> output = layer(input)
+    >>> output.shape
+    torch.Size([10, 2])
+
+    """
+
+    def __init__(self,
+                 activation: str = 'relu',
+                 biases_initializer: str = 'constant_',
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:4360]: Is it possible the bias is causing the error on TF matching? Worth double checking

File: deepchem/models/torch_models/layers.py [hunk 15]
```diff
@@ -4332,3 +4332,112 @@ def gaussian_histogram(self, x: torch.Tensor) -> torch.Tensor:
...
+            will not include biases.
+        weights_initializer: str, optional
+            the initializer for weight values
+
+        """
+        super(Highway, self).__init__(**kwargs)
+        self.activation = activation
+        self.biases_initializer = biases_initializer
+        self.weights_initializer = weights_initializer
+        self.activation_fn = get_activation(self.activation)
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:4380]: Is the activation function different from TF? Worth checking

File: deepchem/models/torch_models/layers.py [hunk 16]
```diff
@@ -4332,3 +4332,112 @@ def gaussian_histogram(self, x: torch.Tensor) -> torch.Tensor:
...
+        init_func = getattr(nn.init, self.weights_initializer)
+        init_func(self.linear_H.weight)
+        init_func(self.linear_T.weight)
+
+        if self.biases_initializer is not None and isinstance(
+                self.biases_initializer, str):
+            bias_init_func = getattr(nn.init, self.biases_initializer)
+            bias_init_func(self.linear_H.bias, -1)
+            bias_init_func(self.linear_T.bias, -1)
+        self.Sigmoid_layer = nn.Sigmoid()
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:4399]: We follow snake_case so this would be `self.sigmoid_layer`

File: deepchem/models/torch_models/layers.py [hunk 17]
```diff
@@ -4332,3 +4332,112 @@ def gaussian_histogram(self, x: torch.Tensor) -> torch.Tensor:
...
+
+        """
+        return (
+            f'{self.__class__.__name__}(activation_fn={self.activation_fn}, '
+            f'biases_initializer={self.biases_initializer}, '
+            f'weights_initializer={self.weights_initializer})')
+
+    def forward(self, inputs: torch.Tensor):
+        """Applies the Highway transformation to the input tensor.
+        Parameters
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:4421]: Spacing here is weird. Make sure to add in spaces

File: deepchem/models/torch_models/tests/highway_test.py [hunk 18]
```diff
@@ -0,0 +1,68 @@
+import torch
+import torch.nn as nn
+import torch.nn.init as init
+import torch.nn.functional as F
+from deepchem.models.torch_models.layers import Highway
+import numpy as np
+
+class Highway(torch.nn.Module):
```
Comment (rbharath) [inline deepchem/models/torch_models/tests/highway_test.py:8]: This a duplicate file, let's remove
